samples/event-sourcing-esdb-simple/src/main/java/io/eventdriven/ecommerce/core/subscriptions/EventStoreDBSubscriptionToAll.java

```diff
@@ -0,0 +1,181 @@
+    try {
+      if (isEventWithEmptyData(resolvedEvent) || isCheckpointEvent(resolvedEvent))
+        return;
+
+      var eventClass = EventTypeMapper.toClass(resolvedEvent.getEvent().getEventType());
+
+      var streamEvent = EventEnvelope.of(eventClass, resolvedEvent);
+
+      if (streamEvent == null) {
+        // That can happen if we're sharing database between modules.
+        // If we're subscribing to all and not filtering out events from other modules,
+        // then we might get events that are from other module and we might not be able to deserialize them.
+        // In that case it's safe to ignore deserialization error.
+        // You may add more sophisticated logic checking if it should be ignored or not.
+        System.out.println("Couldn't deserialize event with id: %s".formatted(resolvedEvent.getEvent().getEventId()));
+
+        if (!subscriptionOptions.ignoreDeserializationErrors())
+          throw new IllegalStateException(
+            "Unable to deserialize event %s with id: %s"
+              .formatted(resolvedEvent.getEvent().getEventType(), resolvedEvent.getEvent().getEventId())
+          );
+
+        return;
+      }
+
+      // publish event to internal event bus
+      eventBus.publish(eventClass, streamEvent);
+
+      checkpointRepository.store(
+        this.subscriptionOptions.subscriptionId(),
+        resolvedEvent.getEvent().getPosition().getCommitUnsigned()
+      );
+    } catch (Exception e) {
+      System.out.println("Error consuming message: {%s}".formatted(e.getMessage()));
+      e.printStackTrace();
+      // if you're fine with dropping some events instead of stopping subscription
+      // then you can add some logic if error should be ignored
+      throw e;
+    }
+  }
+
+  private void handleDrop(Subscription subscription, Throwable throwable) {
+    System.out.println("Subscription was dropped due to " + throwable.getMessage());
+
+    resubscribe();
+  }
+
+  private void resubscribe() {
+    // You may consider adding a max resubscribe count if you want to fail process
+    // instead of retrying until database is up
+    while (true) {
+      var resubscribed = false;
+      synchronized (resubscribeLock) {
+        try {
+          subscribeToAll(subscriptionOptions);
+
+          resubscribed = true;
+        } catch (Exception exception) {
+          System.out.println("Failed to resubscribe to all '%s' dropped with '%s'".formatted(
+            subscriptionOptions.subscriptionId(), exception.getMessage())
+          );
+
+          exception.printStackTrace();
+        }
+      }
+
+      if (resubscribed)
+        break;
+
+      // Sleep between reconnections to not flood the database or not kill the CPU with infinite loop
+      // Randomness added to reduce the chance of multiple subscriptions trying to reconnect at the same time
+      try {
+        Thread.sleep(1000 + new Random().nextInt(1000));
```

Comment:
Sleeps - very difficult to test. You may use some abstraction for that.

Reply:
I'll be testing that with integration tests. It's critical to have some delay between retries not to flood the database with continued requests. I haven't found a better way to retry policy yet, but I'll try to find a better way, as I'm not fully happy with `Thread.sleep`.

Reply:
I throw away my custom, hand-made resubscribe logic and used [Retry](https://docs.spring.io/spring-batch/docs/current/reference/html/retry.html) from Spring Boot 😃